go/fixchain/fix.go

```diff
@@ -71,45 +72,141 @@ func (fix *toFix) constructChain() ([][]*x509.Certificate, []*FixError) {
+// chain, and uses the cache in the toFix struct to go and get any potentially
+// missing intermediate certs.
+// toFix.fixChain() returns a slice of valid and verified chains for this cert
+// to the roots in the toFix struct, and a slice of the errors encountered
+// during the fixing process.
 func (fix *toFix) fixChain() ([][]*x509.Certificate, []*FixError) {
-	var ferrs []*FixError
-	d := *fix.chain
-	d.addCert(fix.cert)
-	for _, c := range d.certs {
-		urls := c.IssuingCertificateURL
-		for _, url := range urls {
-			ferr := fix.augmentIntermediates(url)
-			if ferr != nil {
-				ferrs = append(ferrs, ferr)
-			}
-			chains, err := fix.cert.Verify(*fix.opts)
-			if err == nil {
-				return chains, nil
+	var retferrs []*FixError
+
+	// Ensure the leaf certificate is included as part of the certificate chain.
+	dchain := *fix.chain
+	dchain.addCertToFront(fix.cert)
+
+	explored := make([]bool, len(dchain.certs))
+	lookup := make(map[[hashSize]byte]int)
+	for i, cert := range dchain.certs {
+		lookup[hash(cert)] = i
+	}
+
+	// For each certificate in the given certificate chain...
+	for i, cert := range dchain.certs {
+		// If the chains from this certificate have already been built and
+		// added to the pool of intermediates, skip.
+		if explored[i] {
+			continue
+		}
+
+		seen := make(map[[hashSize]byte]bool)
+		// Build all the chains possible that begin from this certificate,
+		// and add each certificate found along the way to the pool of
+		// intermediates against which to verify fix.cert.  If the addition of
+		// these intermediates causes chains for fix.cert to be verified,
+		// fix.augmentIntermediates() will return those chains.
+		chains, ferrs := fix.augmentIntermediates(cert, 1, seen)
+		if ferrs != nil {
+			retferrs = append(retferrs, ferrs...)
+		}
+		// If adding certs from the chains steming from this cert resulted in
+		// successful verification of chains for fix.cert to fix.root, return
+		// the chains.
+		if chains != nil {
+			return chains, retferrs
+		}
+
+		// Mark any seen certs that match certs in the original chain as already
+		// explored.
+		for certHash := range seen {
+			index, ok := lookup[certHash]
+			if ok {
+				explored[index] = true
 			}
 		}
 	}
-	return nil, append(ferrs, &FixError{
+
+	return nil, append(retferrs, &FixError{
 		Type:  FixFailed,
 		Cert:  fix.cert,
 		Chain: fix.chain.certs,
 	})
 }
 
-func (fix *toFix) augmentIntermediates(url string) *FixError {
+// toFix.augmentIntermediates() builds all possible chains that stem from the
+// given cert, and adds every certificate it finds in these chains to the pool
+// of intermediate certs in toFix.opts.  Every time a new certificate is added
+// to this pool, it tries to re-verify toFix.cert wrt toFix.roots.
+// If this verification is ever successful, toFix.augmentIntermediates() returns
+// the verified chains for toFix.cert wrt toFix.roots.  Also returned are any
+// errors that were encountered along the way.
```

Comment:
Because of this, we may actually not build all possible chains for toFix.cert, which may mean we skip some intermediates.

Not in this version, but let's add a TODO to make it try to construct all possible chains and attempt to log the resulting intermediates.